### PR TITLE
FPGA Prototyping GPIO & Freedom SDK Integration

### PR DESCRIPTION
**Pull request by alexdlukens**
**Type of change**: new feature 

**Impact**: software change

**Release Notes**
I recently created a pull request based on the arty-spi-flash branch #905, but I believe that these changes would be better on a separate branch. My improvements are built on top of the arty-spi-flash branch, however.

In my commits, I have improved compatability with utilizing the Freedom E SDK to generate bare-metal code for Chipyard FPGA designs. Specifically I have added GPIO functionality using the Sifive-Blocks project, added config fragments to ensure that the DTS file is generated with all required information, and added a Makefile target & script to generate board-support files in the Freedom E SDK project automatically. 

Separate from this, I have LED device generation working with Chipyard (showing up in the generated DTS file), however these changes required altering the Sifive-Blocks submodule, which I am not sure is a good solution.

### CODE REVIEW

**Comment by alexdlukens** [comment]
It is unclear to me how to request a new branch be created for these changes, however that is my goal with creating this separate pull request

File: fpga/src/main/scala/arty/Configs.scala [hunk 1]
```diff
@@ -25,17 +29,41 @@ class WithDefaultPeripherals extends Config((site, here, up) => {
...
+  new WithArtyGPIOHarnessBinder ++
+  new chipyard.iobinders.WithGPIOIOCells ++
   new WithArtyJTAGHarnessBinder ++
   new WithArtyUARTHarnessBinder ++
   new WithArtyResetHarnessBinder ++
+  new WithArtySPIFlashHarnessBinder ++
+  new WithSPIFlashIOPassthrough ++
   new WithDebugResetPassthrough ++
   new WithDefaultPeripherals ++
+  new chipyard.config.WithPeripheryBusFrequency(32.0) ++
```
Comment (alexdlukens) [inline fpga/src/main/scala/arty/Configs.scala:56]: This line is crucial to Freedom SDK support, as it tells Freedom what frequency the bus is run at (assumes 100MHz if this is not set). Without this line, interfacing to the board using UART gives garbage data/wrong baud rate

**Comment by alonamid** [comment]
Could you please clarify why you believe that these changes would be better sitting on a separate branch, and what would be the advantage of a branch vs. a fork?

**Comment by alexdlukens** [comment]
I believe the Freedom SDK workflow is very attractive to those using Chipyard FPGA prototyping, and adding this integration/functionality will help improve the ease of utilizing bitstreams generated in Chipyard. As for why this belongs in a separate branch, I am unsure what other improvements @jamesdunn wanted to implement in the arty-spi-flash branch, but my changes are separate from just the spi-flash. I initially PR'd there because my changes were based on-top of the arty-spi-flash branch

I'm currently working on a forked repo so I'm not quite sure what you mean there

**Comment by alonamid** [comment]
Could you provide some more details about which features the Freedom SDK provides that are currently not supported within the Chipyard bare-metal software flows?

**Comment by alexdlukens** [comment]
The Freedom SDK uses a DTS file to generate a header file for use with the [Freedom Metal C library](https://github.com/sifive/freedom-metal) to access peripheral devices laid out in the device tree. It also provides make targets for compiling C programs from a target folder, provides custom configurations for using OpenOCD to upload a program to the fpga via JTAG, and starting a debugging session using GDB. 

I see documentation in section 8.2 of the Chipyard docs for interfacing to a simulation via JTAG, but I am not aware if this works for fpga targets. If these features already exist in Chipyard, please do let me know, I am unaware.

**Comment by jamesdunn** [comment]
Hi @alexdlukens ,

Thank you for your work on this and I apologize for being so late on your review request in the other PR. I replied there and think that the GPIO work looks good and can be merged into arty-spi-flash (since that branch also fixes the UART issue, it's perfectly reasonable to be a place for other IO work). My only suggestion is that the GPIO live in WithArtyTweaks rather than being its own Config and SUB_PROJECT. I will make those suggestions in that PR tomorrow. Once I resolve the arty-sim issue, I would like to simulate arty-spi-flash and merge both into dev or master.

I like your ideas for integrating freedom-e-sdk and metal support into the arty flow, since those are very nice things for embedded development, which is what many are going to use the arty target for. I think those would be best suited for a separate branch that we could discuss. I would want to think through the toolchain recompilation, environment modification and bsp generation scripts separately, as well as think about organizational questions. For example, whether we want to support other targets of freedom-e-sdk like spike in chipyard, and if such support should be more generalized.

I propose we merge your other PR with just the GPIO work into arty-spi-flash. Then I will create a branch from there for the freedom-e-sdk and metal support, which you can PR your scripts into and we can have a discussion on those things.

Thanks,
James

**Comment by alexdlukens** [comment]
Thank you for the response @jamesdunn,

I will move forward with putting GPIO related commits into the PR onthe arty-spi-flash branch, and halt on work for this PR for the moment. I agree that any work done towards the freedom-e-sdk should be on top of your work in the arty-spi-flash branch and that more consideration should be put into generalizing support. Should this PR be closed then?

**Comment by jamesdunn** [comment]
That sounds good, @alexdlukens. I will close this PR and continue the reviews in the other PR for the GPIO work. After that, I will create another branch from there for freedom integration work that you can submit PRs to.

**Pull request closed without merging**